voxblox/src/integrator/tsdf_integrator.cc

```diff
@@ -0,0 +1,567 @@
+  // by a factor, we prevent to easily fade out these areas with the free
+  // space parts of other rays which pass through the corresponding voxels.
+  // This can be useful for creating a TSDF map from sparse sensor data (e.g.
+  // visual features from a SLAM system). By default, this option is disabled.
+  if (config_.use_sparsity_compensation_factor) {
+    if (std::abs(sdf) < config_.default_truncation_distance) {
+      updated_weight *= config_.sparsity_compensation_factor;
+    }
+  }
+
+  // Lookup the mutex that is responsible for this voxel and lock it
+  std::lock_guard<std::mutex> lock(
+      mutexes_.get(getGridIndexFromPoint(point_G, voxel_size_inv_)));
+
+  const float new_weight = tsdf_voxel->weight + updated_weight;
+
+  // it is possible to have weights very close to zero, due to the limited
+  // precision of floating points dividing by this small value can cause nans
+  if (new_weight < kFloatEpsilon) {
+    return;
+  }
+
+  const float new_sdf =
+      (sdf * updated_weight + tsdf_voxel->distance * tsdf_voxel->weight) /
+      new_weight;
+
+  // color blending is expensive only do it close to the surface
+  if (std::abs(sdf) < config_.default_truncation_distance) {
+    tsdf_voxel->color = Color::blendTwoColors(
+        tsdf_voxel->color, tsdf_voxel->weight, color, updated_weight);
+  }
+
+  tsdf_voxel->distance =
+      (new_sdf > 0.0) ? std::min(config_.default_truncation_distance, new_sdf)
+                      : std::max(-config_.default_truncation_distance, new_sdf);
+  tsdf_voxel->weight = std::min(config_.max_weight, new_weight);
+}
+
+// Thread safe.
+// Figure out whether the voxel is behind or in front of the surface.
+// To do this, project the voxel_center onto the ray from origin to point G.
+// Then check if the the magnitude of the vector is smaller or greater than
+// the original distance...
+inline float TsdfIntegratorBase::computeDistance(
+    const Point& origin, const Point& point_G,
+    const Point& voxel_center) const {
+  const Point v_voxel_origin = voxel_center - origin;
+  const Point v_point_origin = point_G - origin;
+
+  const FloatingPoint dist_G = v_point_origin.norm();
+  // projection of a (v_voxel_origin) onto b (v_point_origin)
+  const FloatingPoint dist_G_V = v_voxel_origin.dot(v_point_origin) / dist_G;
+
+  const float sdf = static_cast<float>(dist_G - dist_G_V);
+  return sdf;
+}
+
+// Thread safe.
+inline float TsdfIntegratorBase::getVoxelWeight(const Point& point_C) const {
+  if (config_.use_const_weight) {
+    return 1.0f;
+  }
+  const FloatingPoint dist_z = std::abs(point_C.z());
+  if (dist_z > kEpsilon) {
+    return 1.0f / (dist_z * dist_z);
+  }
+  return 0.0f;
+}
+
+void SimpleTsdfIntegrator::integratePointCloud(const Transformation& T_G_C,
+                                               const Pointcloud& points_C,
+                                               const Colors& colors) {
+  std::vector<VoxelMap> temp_voxel_storage(config_.integrator_threads);
```

Comment:
If we want to multi-thread then unless we serialize access to the map, its size cannot change while the threads are working on it. This leaves two choices pre-allocate the blocks before each update or using temporary storage for missed blocks and merging them in later. (Or using locks inside the map but this slows things down a lot).

The pre-allocation of blocks for the cow and lady dataset usually takes 50 to 100 ms per frame. With the hack I was doing of early termination I got this down to 10 to 20 ms, but it still had issues. The cost of copying these missed blocks is usually 5 to 10 ms (much more for the first frame but much less for most of the others). 

It could be improved further, but it doesn't seem to be bottle-necking things.